File: basic_dsa_two_hundred_ninety_two.java
/**
 * Basic DSA Problems - Two Sum II - Input Array Is Sorted
 * Find two indices with sum = target in sorted array
 */
public class basic_dsa_two_hundred_ninety_two {

    // Problem: Sorted array. Find indices (1-indexed) of two numbers that add to target.
    // Exactly one solution exists.

    /**
     * Two pointers: left at start, right at end. If sum < target, left++. Else right--.
     * Time O(n), Space O(1)
     */
    public static int[] twoSum(int[] numbers, int target) {
        int i = 0, j = numbers.length - 1;
        while (i < j) {
            int sum = numbers[i] + numbers[j];
            if (sum == target) return new int[]{i + 1, j + 1};
            if (sum < target) i++;
            else j--;
        }
        return new int[]{};
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Two Sum II - Input Array Is Sorted");
        System.out.println("=".repeat(60));

        System.out.println("[2,7,11,15], target=9 -> " + java.util.Arrays.toString(twoSum(new int[]{2, 7, 11, 15}, 9)));
        System.out.println("[2,3,4], target=6 -> " + java.util.Arrays.toString(twoSum(new int[]{2, 3, 4}, 6)));
    }
}

File: basic_dsa_two_hundred_ninety_two.cpp
/**
 * Basic DSA - Sum of Left Leaves (LeetCode 404)
 * Return sum of all left leaves in binary tree.
 */
#include <iostream>

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int sumOfLeftLeaves(TreeNode* root, bool isLeft = false) {
    if (!root) return 0;
    if (!root->left && !root->right) return isLeft ? root->val : 0;
    return sumOfLeftLeaves(root->left, true) + sumOfLeftLeaves(root->right, false);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Sum of Left Leaves\n";
    std::cout << std::string(60, '=') << "\n";

    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);

    std::cout << "[3,9,20,null,15,7] -> " << sumOfLeftLeaves(root) << "\n";

    TreeNode* root2 = new TreeNode(1);
    std::cout << "[1] -> " << sumOfLeftLeaves(root2) << "\n";
}
